File: auditoria/parser_atendimentos.py
# -*- coding: utf-8 -*-
"""
Protótipo de parser de Atendimentos (Santri ADM - relatório PDF).
Extrai, por atendimento:
  - numero de controle, prioridade, data, solicitante, status final, tipo (transferencia/uso e consumo)
  - texto original do pedido (a parte "Aberto" mais antiga do tramite)
  - sub-pedidos: quando o atendimento cita mais de uma rota (origem->destino) ou mais de um
    numero de PEDIDO, cada um vira um sub-registro com sua propria lista de produtos/quantidades
"""
import re
import sys
import json

# separador de BLOCO (atendimento->atendimento): comeca na coluna 0, sem indentacao.
# separadores DENTRO do bloco (entre eventos do tramite) vem indentados com espacos - nao contam.
DASH_LINE = re.compile(r'^-{20,}$')  # aplicado apenas a linha SEM strip (ver split_blocks)

# linha-resumo: "220.971      Alta     29/08/2026      4.056 - EVERTHON 430 TRANSFERENCIA CFC PARA CFS ( DAR BAIXA NA CAF    Baixado         016.002 - TRANSFERENCIA"
SUMMARY_RE = re.compile(
    r'^\s*(\d{2,3}\.\d{3})\s+(Alta|Baixa|Média|Media)\s+(\d{2}/\d{2}/\d{4})\s+(.*?)\s+'
    r'(Baixado|Cancelado|Homologado|Homologação cancelada|Atendido|Aberto|Pausado|Em atendimento|Aguardando requisição)\s+'
    r'(\d{3}\.\d{3}\s*-\s*.+)$'
)

# marcador "Aberto  DATA  HORA  SOLICITANTE" (linha de abertura original dentro do historico)
ABERTO_RE = re.compile(
    r'^\s*Aberto\s+(\d{2}/\d{2}/\d{4})\s+(\d{2}:\d{2}:\d{2})\s+(.+?)\s*$'
)

# marcadores de rota origem->destino (varias variacoes encontradas no texto real)
ROUTE_PATTERNS = [
    r'\bDA\s+(?P<origem>[A-Z]{2,5})\s+PARA\s+(?:O\s+|A\s+)?(?P<destino>[A-Z]{2,5})\b',
    r'\b(?P<origem>[A-Z]{2,5})\s*[>\-]{1,2}\s*(?P<destino>[A-Z]{2,5})\b',
    r'\b(?P<origem>[A-Z]{2,5})\s+P/\s*(?P<destino>[A-Z]{2,5})\b',
    r'\b(?P<origem>[A-Z]{2,5})\s+PARA\s+(?P<destino>[A-Z]{2,5})\b',
    r'\bLOJA\s+(?P<origem>[A-Z]{2,5})\s+DESTINO\s+(?P<destino>[A-Z]{2,5})\b',
    r'\bDE:\s*(?P<origem>[A-Z]{2,5})\b.*?\bPARA:\s*(?P<destino>[A-Z]{2,5})\b',
]
ROUTE_RES = [re.compile(p, re.IGNORECASE) for p in ROUTE_PATTERNS]

KNOWN_EMPRESAS = {"CFS","CFVP","CFR","CFC","CFW3","CFT","CFG","CFJB","CFPA","CFBS"}

PEDIDO_RE = re.compile(r'PEDIDO[:\s]*([0-9]{6,})', re.IGNORECASE)

# produto + quantidade em varios formatos encontrados no texto real:
#  "4234>6 und", "119927---------7 und", "COD:130232", "ADM 137.853 QUANT 5",
#  "130931 QT 03", "51.188----- 1", "COD:137.997 4 INID", "126922 QT: 17,93 MTS"
PRODUCT_LINE_RE = re.compile(
    r'(?:COD|ADM|CODIGO)?\.?:?\s*'
    r'\(?'                                                  # parenteses opcionais em volta (ex: "68.945(04 UNID.)")
    r'(\d{1,3}(?:\.\d{3})+|\d{4,8})'                      # codigo do produto
    r'\)?'
    r'\s*(?:[-\.>=/]{1,25}|QT\.?D?\.?|QUANT\.?|QUANTIDADE|QDT\.?)?\s*[:=]?\s*'  # separador (simbolo, palavra, ou so espaco)
    r'\(?'
    r'(\d+(?:[.,]\d+)?)\s*'                                # quantidade
    r'(UND\.?|UN\.?|UNID\.?|UNIDADES?|UNI\.?|MTS?|M2|M3|SC|KG|LT|LITROS?|PCT|CX|PE[ÇC]AS?)?'
    r'\)?',
    re.IGNORECASE
)

def norm_code(s):
    return s.replace('.', '')

FURNITURE_EXACT = {
    'Status','Baixado','Homologado','Homologação cancelada','Em homologação',
    'Atendido','Pausado','Em atendimento','Aguardando requisição','Aberto',
}

def is_furniture(line):
    s = line.strip()
    if not s:
        return True
    if DASH_LINE.match(s):                       # qualquer linha de tracejado (titulo OU separador de atend.)
        return True
    if s.startswith('Acompanhamento de Controles'):
        return True
    if s.startswith('ADM 1.1.5.9'):               # rodape "ADM 1.1.5.9 R1 Emitido em ... Página NNNN/0042"
        return True
    if 'F I L T R O S' in s or s.startswith('Ordenação'):
        return True
    if s.startswith('Controle atend.') and 'Tipo de controle' in s:  # cabecalho de coluna (repete a cada pagina)
        return True
    if s in FURNITURE_EXACT:                      # legenda de status isolada (bloco de filtros no fim do doc)
        return True
    return False

def split_blocks(raw_text):
    """
    Estrategia robusta a quebra de pagina: remove toda a "mobilia" do relatorio
    (titulo, cabecalho de coluna repetido a cada pagina, rodape, form-feed, linhas
    de tracejado) e depois usa a propria linha-resumo (SUMMARY_RE) como ancora de
    inicio de cada atendimento — assim, se o texto original de um atendimento for
    longo o suficiente pra atravessar uma quebra de pagina, ele continua sendo
    tratado como UM UNICO bloco (a mobilia no meio e' descartada, nao quebra o bloco).
    """
    raw_text = raw_text.replace('\f', '\n')
    lines = [l for l in raw_text.split('\n') if not is_furniture(l)]

    blocks = []
    current = []
    for line in lines:
        if SUMMARY_RE.match(line):
            if current:
                blocks.append(current)
            current = [line]
        else:
            if current:  # ignora qualquer coisa antes do primeiro resumo encontrado
                current.append(line)
    if current:
        blocks.append(current)
    return blocks

def find_summary(block_lines):
    for line in block_lines:
        m = SUMMARY_RE.match(line)
        if m:
            return {
                'controle': m.group(1),
                'prioridade': m.group(2),
                'data': m.group(3),
                'solicitante_assunto_raw': m.group(4).strip(),
                'status_final': m.group(5),
                'tipo_controle': m.group(6).strip(),
            }
    return None

def find_original_request_text(block_lines):
    """
    O texto original do pedido fica ENTRE a ultima linha 'Aberto DATA HORA SOLICITANTE'
    (mais antiga, no fim do bloco no layout deste relatorio) e o fim do bloco.
    """
    aberto_idx = None
    for i, line in enumerate(block_lines):
        if ABERTO_RE.match(line):
            aberto_idx = i  # mantem a ULTIMA ocorrencia (a mais antiga = pedido original)
    if aberto_idx is None:
        return ''
    text_lines = block_lines[aberto_idx+1:]
    text_lines = [l.strip() for l in text_lines if l.strip()]
    return '\n'.join(text_lines)

def extract_pedidos_numbers(text):
    return PEDIDO_RE.findall(text)

def extract_route_segments(text):
    """
    Divide o texto em segmentos por marcador de rota. Retorna lista de dicts:
    {origem, destino, texto_produtos}
    Se nao achar nenhuma rota, retorna 1 segmento sem origem/destino (rota indefinida).
    """
    matches = []
    for rx in ROUTE_RES:
        for m in rx.finditer(text):
            o, d = m.group('origem').upper(), m.group('destino').upper()
            if o in KNOWN_EMPRESAS and d in KNOWN_EMPRESAS:
                matches.append((m.start(), m.end(), o, d))
    matches.sort(key=lambda x: x[0])
    # dedup por posicao proxima (regex sobrepostas)
    dedup = []
    for m in matches:
        if dedup and m[0] - dedup[-1][1] < 3:
            continue
        dedup.append(m)

    if not dedup:
        return [{'origem': None, 'destino': None, 'texto': text}]

    segments = []
    for i, (start, end, o, d) in enumerate(dedup):
        seg_start = end
        seg_end = dedup[i+1][0] if i+1 < len(dedup) else len(text)
        segments.append({'origem': o, 'destino': d, 'texto': text[seg_start:seg_end].strip()})
    # texto antes da primeira rota (pode conter produtos que pertencem a ela, dependendo do padrao)
    pre_text = text[:dedup[0][0]].strip()
    if pre_text:
        segments[0]['texto'] = pre_text + '\n' + segments[0]['texto']
    return segments

def extract_products(text):
    # remove linhas de "PEDIDO: NNNNN" antes de procurar produtos, senao o proprio
    # numero do pedido de cliente e capturado como se fosse codigo de produto
    text_sem_pedido = PEDIDO_RE.sub('', text)
    produtos = []
    for m in PRODUCT_LINE_RE.finditer(text_sem_pedido):
        codigo, qtd, und = m.groups()
        produtos.append({
            'codigo': norm_code(codigo),
            'quantidade': qtd.replace(',', '.'),
            'unidade': (und or '').upper() or None,
        })
    return produtos

def parse_atendimento_block(block_lines):
    summary = find_summary(block_lines)
    if not summary:
        return None
    original_text = find_original_request_text(block_lines)
    pedidos = extract_pedidos_numbers(original_text)
    segments = extract_route_segments(original_text)

    sub_pedidos = []
    for seg in segments:
        produtos = extract_products(seg['texto'])
        seg_pedidos = extract_pedidos_numbers(seg['texto'])
        sub_pedidos.append({
            'origem': seg['origem'],
            'destino': seg['destino'],
            'pedido_cliente': seg_pedidos[0] if seg_pedidos else None,
            'produtos': produtos,
            'texto_bruto': seg['texto'][:300],
        })

    tem_produto = any(sp['produtos'] for sp in sub_pedidos)

    return {
        'controle': summary['controle'],
        'data': summary['data'],
        'status_final': summary['status_final'],
        'tipo_controle': summary['tipo_controle'],
        'tem_pedido_cliente': len(pedidos) > 0,
        'pedidos_clientes': pedidos,
        'qtd_sub_rotas': len([s for s in sub_pedidos if s['origem']]),
        'sub_pedidos': sub_pedidos,
        'texto_original': original_text[:500],
        # marca se o atendimento entra na auditoria automatica ou precisa de revisao manual
        'status_extracao': 'ok' if tem_produto else 'revisao_manual',
        'motivo_revisao': None if tem_produto else 'nenhum produto/quantidade identificado no texto do pedido',
    }

def parse_file(path):
    """
    Retorna (auditaveis, revisao_manual, blocos_sem_resumo)
      auditaveis     -> atendimentos com pelo menos 1 produto extraido, prontos pro cruzamento
      revisao_manual -> atendimentos parseados mas sem produto identificado (isolados, nao entram
                         no cruzamento automatico ate alguem confirmar manualmente o conteudo)
      blocos_sem_resumo -> nao foi nem possivel identificar o resumo do atendimento (corrupcao no PDF fonte)
    """
    with open(path, encoding='utf-8') as f:
        raw = f.read()
    blocks = split_blocks(raw)
    auditaveis, revisao_manual = [], []
    sem_resumo = 0
    for b in blocks:
        parsed = parse_atendimento_block(b)
        if parsed is None:
            joined = '\n'.join(b).strip()
            if joined:
                sem_resumo += 1
            continue
        if parsed['status_extracao'] == 'ok':
            auditaveis.append(parsed)
        else:
            revisao_manual.append(parsed)
    return auditaveis, revisao_manual, sem_resumo

if __name__ == '__main__':
    path = sys.argv[1]
    auditaveis, revisao_manual, sem_resumo = parse_file(path)
    print(f"Auditaveis (com produto): {len(auditaveis)}")
    print(f"Revisao manual (sem produto): {len(revisao_manual)}")
    print(f"Corrompidos no PDF fonte (sem resumo identificavel): {sem_resumo}")

File: parser_atendimentos.js
// Porta fiel de parser_atendimentos.py para JavaScript (roda no navegador).
// Qualquer mudanca aqui deve ser espelhada la, e vice-versa.

const DASH_LINE = /^-{20,}$/;

const SUMMARY_RE = new RegExp(
  '^\\s*(\\d{2,3}\\.\\d{3})\\s+(Alta|Baixa|Média|Media)\\s+(\\d{2}/\\d{2}/\\d{4})\\s+(.*?)\\s+' +
  '(Baixado|Cancelado|Homologado|Homologação cancelada|Atendido|Aberto|Pausado|Em atendimento|Aguardando requisição)\\s+' +
  '(\\d{3}\\.\\d{3}\\s*-\\s*.+)$'
);

const ABERTO_RE = /^\s*Aberto\s+(\d{2}\/\d{2}\/\d{4})\s+(\d{2}:\d{2}:\d{2})\s+(.+?)\s*$/;

const FURNITURE_EXACT = new Set([
  'Status','Baixado','Homologado','Homologação cancelada','Em homologação',
  'Atendido','Pausado','Em atendimento','Aguardando requisição','Aberto',
]);

function isFurniture(line){
  const s = line.trim();
  if (!s) return true;
  if (DASH_LINE.test(s)) return true;
  if (s.startsWith('Acompanhamento de Controles')) return true;
  if (s.startsWith('ADM 1.1.5.9')) return true;
  if (s.includes('F I L T R O S') || s.startsWith('Ordenação')) return true;
  if (s.startsWith('Controle atend.') && s.includes('Tipo de controle')) return true;
  if (FURNITURE_EXACT.has(s)) return true;
  return false;
}

function splitBlocks(rawText){
  rawText = rawText.split('\f').join('\n');
  const lines = rawText.split('\n').filter(l => !isFurniture(l));
  const blocks = [];
  let current = [];
  for (const line of lines){
    if (SUMMARY_RE.test(line)){
      if (current.length) blocks.push(current);
      current = [line];
    } else if (current.length) {
      current.push(line);
    }
  }
  if (current.length) blocks.push(current);
  return blocks;
}

function findSummary(blockLines){
  for (const line of blockLines){
    const m = SUMMARY_RE.exec(line);
    if (m){
      return {
        controle: m[1], prioridade: m[2], data: m[3],
        solicitante_assunto_raw: m[4].trim(), status_final: m[5],
        tipo_controle: m[6].trim(),
      };
    }
  }
  return null;
}

function findUsuarioBaixa(blockLines){
  const re = /^\s*Baixado\s+\d{2}\/\d{2}\/\d{4}\s+\d{2}:\d{2}:\d{2}\s+(.+?)\s*$/;
  for (const line of blockLines){
    const m = re.exec(line);
    if (m) return m[1].trim();
  }
  return null;
}

function findOriginalRequestText(blockLines){
  let abertoIdx = null;
  for (let i = 0; i < blockLines.length; i++){
    if (ABERTO_RE.test(blockLines[i])) abertoIdx = i;
  }
  if (abertoIdx === null) return '';
  return blockLines.slice(abertoIdx + 1).map(l => l.trim()).filter(Boolean).join('\n');
}

const PEDIDO_RE = /PEDIDO[:\s]*([0-9]{6,})/gi;

function extractPedidosNumbers(text){
  const out = [];
  let m;
  const re = new RegExp(PEDIDO_RE.source, 'gi');
  while ((m = re.exec(text)) !== null) out.push(m[1]);
  return out;
}

// ---------------- lojas / rotas ----------------
const KNOWN_EMPRESAS = new Set(["CFS","CFVP","CFR","CFC","CFW3","CFT","CFG","CFJB","CFPA","CFBS","CAP99"]);

const ALIASES_LOJA = {
  CFS:   ["CFS", "SAMAMBAIA"],
  CFR:   ["CFR", "RECANTO"],
  CFVP:  ["CFVP", "VICENTE PIRES", "VICENTE"],
  CFC:   ["CFC", "CEILANDIA"],
  CFW3:  ["CFW3", "ASA NORTE", "W3"],
  CFT:   ["CFT", "TAGUATINGA"],
  CFG:   ["CFG", "GAMA"],
  CFJB:  ["CFJB", "JARDIM BOTANICO"],
  CFPA:  ["CFPA", "PONTE ALTA"],
  CFBS:  ["CFBS", "BERNARDO SAYAO", "BERNADO SAYAO"],
  CAP99: ["CAP99", "CAPITAL ATACADISTA", "CAPITAL"],
};

const ACCENT_MAP = {
  'á':'a','à':'a','â':'a','ã':'a','ä':'a','é':'e','è':'e','ê':'e','ë':'e',
  'í':'i','ì':'i','î':'i','ï':'i','ó':'o','ò':'o','ô':'o','õ':'o','ö':'o',
  'ú':'u','ù':'u','û':'u','ü':'u','ç':'c',
  'Á':'A','À':'A','Â':'A','Ã':'A','Ä':'A','É':'E','È':'E','Ê':'E','Ë':'E',
  'Í':'I','Ì':'I','Î':'I','Ï':'I','Ó':'O','Ò':'O','Ô':'O','Õ':'O','Ö':'O',
  'Ú':'U','Ù':'U','Û':'U','Ü':'U','Ç':'C',
};

function semAcentoMaiuscula(s){
  return s.split('').map(c => ACCENT_MAP[c] || c).join('').toUpperCase();
}

const ALIAS_PARA_CODIGO = {};
for (const [codigo, nomes] of Object.entries(ALIASES_LOJA)){
  for (const nome of nomes) ALIAS_PARA_CODIGO[semAcentoMaiuscula(nome)] = codigo;
}

function resolveEmpresa(texto){
  if (!texto) return null;
  return ALIAS_PARA_CODIGO[semAcentoMaiuscula(texto.trim())] || null;
}

function escapeRegex(s){ return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); }

const TODOS_ALIASES = Object.keys(ALIAS_PARA_CODIGO).sort((a,b) => b.length - a.length);
const PLACE_ALT = TODOS_ALIASES.map(a => escapeRegex(a).replace(/\\ /g, '\\s+')).join('|');
const PLACE_RE_TXT = '(?:' + PLACE_ALT + ')';

const ROUTE_PATTERNS = [
  '\\bDA\\s+(?<origem>' + PLACE_RE_TXT + ')\\s+PARA\\s+(?:O\\s+|A\\s+)?(?<destino>' + PLACE_RE_TXT + ')\\b',
  '\\b(?<origem>' + PLACE_RE_TXT + ')\\s*[>\\-]{1,2}\\s*(?<destino>' + PLACE_RE_TXT + ')\\b',
  '\\b(?<origem>' + PLACE_RE_TXT + ')\\s+P\\/\\s*(?<destino>' + PLACE_RE_TXT + ')\\b',
  '\\b(?<origem>' + PLACE_RE_TXT + ')\\s+PARA\\s+(?<destino>' + PLACE_RE_TXT + ')\\b',
  '\\bLOJA\\s+(?<origem>' + PLACE_RE_TXT + ')\\s+DESTINO\\s+(?<destino>' + PLACE_RE_TXT + ')\\b',
  '\\bDE:\\s*(?<origem>' + PLACE_RE_TXT + ')\\b.*?\\bPARA:\\s*(?<destino>' + PLACE_RE_TXT + ')\\b',
];
const ROUTE_RES = ROUTE_PATTERNS.map(p => new RegExp(p, 'gi'));

const ORIGIN_ONLY_PATTERNS = [
  '\\bDA\\s+LOJA\\s+(?:DE\\s+|DA\\s+|DO\\s+)?(?<origem>' + PLACE_RE_TXT + ')\\b',
  '\\bDA\\s+LOJA\\s+(?<origem>' + PLACE_RE_TXT + ')\\b',
];
const ORIGIN_ONLY_RES = ORIGIN_ONLY_PATTERNS.map(p => new RegExp(p, 'gi'));

function dedupOverlaps(matches){
  const dedup = [];
  for (const m of matches){
    if (dedup.length && m[0] < dedup[dedup.length-1][1]) continue;
    dedup.push(m);
  }
  return dedup;
}

function findRouteMatches(text){
  const textoNorm = semAcentoMaiuscula(text);
  let matches = [];
  for (const rx of ROUTE_RES){
    rx.lastIndex = 0;
    let m;
    while ((m = rx.exec(textoNorm)) !== null){
      const o = resolveEmpresa(m.groups.origem), d = resolveEmpresa(m.groups.destino);
      if (o && d) matches.push([m.index, m.index + m[0].length, o, d]);
      if (m[0].length === 0) rx.lastIndex++;
    }
  }
  for (const rx of ORIGIN_ONLY_RES){
    rx.lastIndex = 0;
    let m;
    while ((m = rx.exec(textoNorm)) !== null){
      const o = resolveEmpresa(m.groups.origem);
      if (o) matches.push([m.index, m.index + m[0].length, o, null]);
      if (m[0].length === 0) rx.lastIndex++;
    }
  }
  matches.sort((a,b) => a[0] - b[0] || (b[1]-b[0]) - (a[1]-a[0]));
  return dedupOverlaps(matches);
}

function segmentsForward(text, matches){
  const segments = [];
  for (let i = 0; i < matches.length; i++){
    const [start, end, o, d] = matches[i];
    const segStart = end;
    const segEnd = i+1 < matches.length ? matches[i+1][0] : text.length;
    segments.push({ origem: o, destino: d, texto: text.slice(segStart, segEnd).trim() });
  }
  const preText = text.slice(0, matches[0][0]).trim();
  if (preText) segments[0].texto = preText + '\n' + segments[0].texto;
  return segments;
}

function segmentsBackward(text, matches){
  const segments = [];
  let prevEnd = 0;
  for (const [start, end, o, d] of matches){
    segments.push({ origem: o, destino: d, texto: text.slice(prevEnd, start).trim() });
    prevEnd = end;
  }
  const trailing = text.slice(prevEnd).trim();
  if (trailing && segments.length) segments[segments.length-1].texto = (segments[segments.length-1].texto + '\n' + trailing).trim();
  return segments;
}

function extractRouteSegments(text){
  const matches = findRouteMatches(text);
  if (!matches.length) return [{ origem: null, destino: null, texto: text }];
  if (matches.length === 1) return segmentsForward(text, matches);

  const forward = segmentsForward(text, matches);
  const backward = segmentsBackward(text, matches);
  const vazios = segs => segs.filter(s => s.origem && !extractProducts(s.texto).length).length;
  return vazios(forward) <= vazios(backward) ? forward : backward;
}

// ---------------- produtos ----------------
function normCode(s){ return s.split('.').join(''); }

const PRODUCT_LINE_RE = new RegExp(
  '(?:COD|ADM|CODIGO)?\\.?:?\\s*' +
  '\\(?' +
  '(?<!\\d)(\\d{1,3}(?:\\.\\d{3})+|\\d{4,6})(?!\\d)' +
  '\\)?' +
  '\\s*(?:[-.>=\\/]{1,25}|QT\\.?D?\\.?|QUANT\\.?|QUANTIDADE|QDT\\.?|QUAT\\.?)?\\s*[:=]?\\s*' +
  '\\(?\\s*' +
  '(\\d+(?:[.,]\\d+)?)\\s*' +
  '(UND\\.?|UN\\.?|UNID\\.?|UNIDADES?|UNI\\.?|MTS?|M2|M3|SC|KG|LT|LITROS?|PCT|CX|PE[ÇC]AS?)?' +
  '\\s*\\)?',
  'gi'
);

const NOISE_LINES = /^(BOM DIA|BOA TARDE|BOA NOITE|OBRIGAD[AO]|GRATO|GRATA|ATT\.?!?|ATENCIOSAMENTE|OBS\.?:?.*|OBSERVA[ÇC][AÃ]O:?.*|SOLICITO A RETIRADA.*|POR FAVOR.*)$/i;

const QTY_FIRST_RE = /^\s*(\d{1,2})\s+([A-Za-zÀ-ÿ][A-Za-zÀ-ÿ0-9"'\s]{2,55}?)(?:\s+(?<!\d)(\d{1,3}(?:\.\d{3})+|\d{4,6})(?!\d))?\s*$/;
const CODE_EQ_NAME_EQ_QTY_RE = /(?<!\d)(\d{1,3}(?:\.\d{3})+|\d{4,6})(?!\d)\s*=\s*([^=\n]{2,55}?)\s*=\s*(\d+(?:[.,]\d+)?)/;
const SHORT_CODE_RE = /^\s*(\d{2,3})\s*>\s*(\d+(?:[.,]\d+)?)\s*(UND\.?|UNIDADES?|UN\.?)?\s*$/i;
const TRAILING_QTY_RE = /([A-ZÀ-Úa-zà-ú][A-Za-zÀ-ÿ]*(?:\s+[A-ZÀ-Úa-zà-ú][A-Za-zÀ-ÿ]*){0,2})\s+(\d{1,3}(?:[.,]\d+)?)\s+(UNIDADES?|UND\.?|UN\.?)\b/gi;
const LEADING_QTY_NAME_RE = /\b(\d{1,2})\s+([A-ZÀ-Ú]{3,}(?:\s+[A-ZÀ-Ú]{2,}){0,2})\s*$/gm;

// produto com nome longo no meio: "CODIGO  NOME DO PRODUTO POR EXTENSO (pode ter
// dimensao tipo 62X62 no meio)  [CX]  QUANTIDADE  UNIDADE" - a quantidade+unidade
// sempre no FINAL da linha. Cobre piso/porcelanato/revestimento (M2/MT) e qualquer
// outro produto com nome/descricao comprida antes da quantidade real (ex:
// "136.197 TELHA RESIDENCIAL 3,66 X 1,10M 6M MULT 01 UN").
const CODE_NOME_QTD_RE = /^\s*(?<!\d)(\d{1,3}(?:\.\d{3})+|\d{4,6})(?!\d)\s+.+?\s+(?:CX\s+)?(\d+(?:[.,]\d+)?)\s*(UND\.?|UN\.?|UNID\.?|UNIDADES?|UNI\.?|MTS?|M2|M3|SC|KG|LT|LITROS?|PCT|PE[ÇC]AS?)\.?\s*$/i;

function extractProductsCodigoNomeQtd(text){
  const produtos = [];
  for (let linha of text.split('\n')){
    linha = linha.trim();
    if (!linha) continue;
    const m = CODE_NOME_QTD_RE.exec(linha);
    if (m){
      produtos.push({ codigo: normCode(m[1]), nome: null, quantidade: m[2].replace(',', '.'), unidade: m[3].toUpperCase(), metodo: 'codigo_nome_qtd' });
    }
  }
  return produtos;
}

function extractProductsFallback(text){
  let produtos = [];
  for (let linha of text.split('\n')){
    linha = linha.trim();
    if (!linha || NOISE_LINES.test(linha)) continue;

    let m = CODE_EQ_NAME_EQ_QTY_RE.exec(linha);
    if (m){
      produtos.push({ codigo: normCode(m[1]), nome: m[2].trim(), quantidade: m[3].replace(',', '.'), unidade: null, metodo: 'codigo=nome=qtd' });
      continue;
    }
    m = QTY_FIRST_RE.exec(linha);
    if (m){
      produtos.push({ codigo: m[3] ? normCode(m[3]) : null, nome: m[2].trim(), quantidade: m[1].replace(',', '.'), unidade: null, metodo: 'qtd_primeiro' });
      continue;
    }
    m = SHORT_CODE_RE.exec(linha);
    if (m){
      produtos.push({ codigo: normCode(m[1]), nome: null, quantidade: m[2].replace(',', '.'), unidade: (m[3]||'').toUpperCase() || null, metodo: 'codigo_curto' });
      continue;
    }
  }
  if (!produtos.length){
    const re1 = new RegExp(TRAILING_QTY_RE.source, 'gi');
    let m;
    while ((m = re1.exec(text)) !== null){
      const nome = m[1].trim();
      if (NOISE_LINES.test(nome) || nome.length < 3) continue;
      produtos.push({ codigo: null, nome, quantidade: m[2].replace(',', '.'), unidade: m[3].toUpperCase(), metodo: 'nome+qtd_final' });
    }
    if (!produtos.length){
      const re2 = new RegExp(LEADING_QTY_NAME_RE.source, 'gm');
      while ((m = re2.exec(text)) !== null){
        const nome = m[2];
        if (NOISE_LINES.test(nome)) continue;
        produtos.push({ codigo: null, nome: nome.trim().replace(/\w\S*/g, t => t.charAt(0).toUpperCase()+t.substr(1).toLowerCase()), quantidade: m[1], unidade: null, metodo: 'qtd+nome_maiusculo' });
      }
    }
  }
  return produtos;
}

function extractProducts(text){
  const textSemPedido = text.replace(new RegExp(PEDIDO_RE.source, 'gi'), '');
  const produtos = [];
  const re = new RegExp(PRODUCT_LINE_RE.source, 'gi');
  let m;
  while ((m = re.exec(textSemPedido)) !== null){
    produtos.push({ codigo: normCode(m[1]), nome: null, quantidade: m[2].replace(',', '.'), unidade: (m[3]||'').toUpperCase() || null, metodo: 'codigo>qtd' });
    if (m[0].length === 0) re.lastIndex++;
  }
  if (!produtos.length){
    const piso = extractProductsCodigoNomeQtd(textSemPedido);
    return piso.length ? piso : extractProductsFallback(textSemPedido);
  }
  return produtos;
}

function parseAtendimentoBlock(blockLines){
  const summary = findSummary(blockLines);
  if (!summary) return null;
  const originalText = findOriginalRequestText(blockLines);
  const pedidos = extractPedidosNumbers(originalText);
  const segments = extractRouteSegments(originalText);

  const subPedidos = segments.map(seg => {
    const produtos = extractProducts(seg.texto);
    const segPedidos = extractPedidosNumbers(seg.texto);
    return {
      origem: seg.origem, destino: seg.destino,
      pedido_cliente: segPedidos.length ? segPedidos[0] : null,
      produtos,
      texto_bruto: seg.texto.slice(0, 300),
    };
  });

  const temProduto = subPedidos.some(sp => sp.produtos.length);

  return {
    controle: summary.controle, data: summary.data, status_final: summary.status_final,
    tipo_controle: summary.tipo_controle, usuario_baixa: findUsuarioBaixa(blockLines),
    tem_pedido_cliente: pedidos.length > 0, pedidos_clientes: pedidos,
    qtd_sub_rotas: subPedidos.filter(s => s.origem).length,
    sub_pedidos: subPedidos, texto_original: originalText.slice(0, 500),
    status_extracao: temProduto ? 'ok' : 'revisao_manual',
    motivo_revisao: temProduto ? null : 'nenhum produto/quantidade identificado no texto do pedido',
  };
}

function parseAtendimentosTexto(rawText){
  const blocks = splitBlocks(rawText);
  const auditaveis = [], revisaoManual = [];
  let semResumo = 0;
  for (const b of blocks){
    const parsed = parseAtendimentoBlock(b);
    if (!parsed){
      if (b.join('\n').trim()) semResumo++;
      continue;
    }
    if (parsed.status_extracao === 'ok') auditaveis.push(parsed);
    else revisaoManual.push(parsed);
  }
  return { auditaveis, revisaoManual, semResumo };
}

